homeassistant/components/habitica.py

```diff
@@ -0,0 +1,92 @@
+"""
+The Habitica API component.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/habitica/
+"""
+
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+from homeassistant.const import \
+   CONF_NAME, CONF_URL
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers import config_validation as cv
+
+REQUIREMENTS = ['habitipy==0.2.0']
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = "habitica"
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=15)
+CONF_API_USER = "api_user"
+CONF_API_KEY = "api_key"
+ATTR_NAME = CONF_NAME
+ATTR_PATH = "path"
+ATTR_ARGS = "args"
+INSTANCE_SCHEMA = vol.Schema({
+    vol.Optional(CONF_URL): cv.url,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_API_USER): cv.string,
+    vol.Required(CONF_API_KEY): cv.string
+})
+
+
+INSTANCE_LIST_SCHEMA = vol.All(cv.ensure_list, [INSTANCE_SCHEMA])
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: INSTANCE_LIST_SCHEMA
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_API_CALL = 'api_call'
+SERVICE_API_CALL_SCHEMA = vol.Schema({
+    vol.Required(ATTR_NAME): str,
+    vol.Required(ATTR_PATH): vol.All(cv.ensure_list, [str]),
+    vol.Optional(ATTR_ARGS): dict
+})
+
+
+async def async_setup(hass, config):
+    """Set up the habitica service."""
+    conf = config.get(DOMAIN, None)
+    data = hass.data[DOMAIN] = {}
+    websession = async_get_clientsession(hass)
+    from habitipy.aio import HabitipyAsync
+
+    class HAHabitipyAsync(HabitipyAsync):
+        """Closure API class to hold session."""
+
+        def __call__(self, **kwargs):
+            return super().__call__(websession, **kwargs)
+
+    for instance in conf:
+        url = instance.get(CONF_URL, 'https://habitica.com')
+        username = instance.get(CONF_API_USER, None)
```

Comment:
config schema will guarantee CONF_API_USER exist, `uername = instance[CONF_API_USER]` is enough.